extensions/standard-processors/processors/GenerateFlowFile.cpp

```diff
@@ -58,57 +58,69 @@ void generateData(std::vector<char>& data, bool textData = false) {
 }
 
 void GenerateFlowFile::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
-  if (context.getProperty(FileSize.name, fileSize_)) {
-    logger_->log_trace("File size is configured to be {}", fileSize_);
+  if (context.getProperty(FileSize.name, file_size_)) {
+    logger_->log_trace("File size is configured to be {}", file_size_);
   }
 
-  if (context.getProperty(BatchSize.name, batchSize_)) {
-    logger_->log_trace("Batch size is configured to be {}", batchSize_);
+  if (context.getProperty(BatchSize.name, batch_size_)) {
+    logger_->log_trace("Batch size is configured to be {}", batch_size_);
   }
 
-  std::string value;
-  if (context.getProperty(DataFormat.name, value)) {
-    textData_ = (value == GenerateFlowFile::DATA_FORMAT_TEXT);
+  if (auto data_format = context.getProperty(DataFormat)) {
+    textData_ = (*data_format == DATA_FORMAT_TEXT);
   }
-  if (context.getProperty(UniqueFlowFiles.name, uniqueFlowFile_)) {
-    logger_->log_trace("Unique Flow files is configured to be {}", uniqueFlowFile_);
+  if (context.getProperty(UniqueFlowFiles.name, unique_flow_file_)) {
+    logger_->log_trace("Unique Flow files is configured to be {}", unique_flow_file_);
   }
 
   std::string custom_text;
   context.getProperty(CustomText, custom_text, nullptr);
   if (!custom_text.empty()) {
-    if (textData_ && !uniqueFlowFile_) {
-      data_.assign(custom_text.begin(), custom_text.end());
+    if (textData_ && !unique_flow_file_) {
+      non_unique_data_.assign(custom_text.begin(), custom_text.end());
       return;
-    } else {
-      logger_->log_warn("Custom Text property is set, but not used!");
     }
+    logger_->log_warn("Custom Text property is set, but not used!");
   }
 
-  if (!uniqueFlowFile_) {
-    data_.resize(gsl::narrow<size_t>(fileSize_));
-    generateData(data_, textData_);
+  if (!unique_flow_file_) {
+    non_unique_data_.resize(gsl::narrow<size_t>(file_size_));
+    generateData(non_unique_data_, textData_);
+  }
+}
+
+// If the Data Format is text and if Unique FlowFiles is false, the custom text has to be evaluated once per batch
+void GenerateFlowFile::regenerateNonUniqueData(core::ProcessContext& context) {
+  std::string custom_text;
+  context.getProperty(CustomText, custom_text, nullptr);
+  if (!custom_text.empty()) {
+    if (textData_ && !unique_flow_file_) {
+      non_unique_data_.assign(custom_text.begin(), custom_text.end());
+      return;
+    }
+    logger_->log_warn("Custom Text property is set, but not used!");
```

Comment:
I think this would be simpler, and would avoid doing some unnecessary work:
```c++
  if (!textData_ || unique_flow_file_) return;

  std::string custom_text;
  context.getProperty(CustomText, custom_text, nullptr);
  if (!custom_text.empty()) {
    non_unique_data_.assign(custom_text.begin(), custom_text.end());
  }
```
We would lose the warning, but we have warned in `onSchedule()` already.

Also, if the expanded custom text is empty now, but it wasn't empty before, I think the flow file should be empty, so
```c++
  if (!textData_ || unique_flow_file_) return;

  std::string custom_text;
  context.getProperty(CustomText, custom_text, nullptr);
  non_unique_data_.assign(custom_text.begin(), custom_text.end());
```
would be better.

Reply:
The last suggestion would remove the possibility to use GenerateFlowFiles with non unique randomly generated set length data. (UniqueFlowFiles -> false, FileSize -> 10B, without CustomData, would generate empty flowfiles)
How about we make a distinction between CustomText not set and CustomText evaluated to be empty?

```
void GenerateFlowFile::regenerateNonUniqueData(core::ProcessContext& context) {
  if (!textData_ || unique_flow_file_) return;

  if (auto custom_text = context.getProperty(CustomText, nullptr)) {
    non_unique_data_.assign(custom_text->begin(), custom_text->end());
  }
}
```

Reply:
Nifi seem to work similarly:
  - CustomText not set -> 10B random text
  - CustomText -> ${something_that_doesnt_exists}

One thing thats different that nifi doesnt allow empty string for customtext

Reply:
https://github.com/apache/nifi-minifi-cpp/pull/1706/commits/89c76dc4a13ab07d6a73385f5eda37aafc2dc31b#diff-45136a0df5089a3527479406b275001e601b5a9e02fa3a345be89e18e141b03bR96-R98

I've also fixed the cut off description of Unique FlowFiles https://github.com/apache/nifi-minifi-cpp/pull/1706/commits/89c76dc4a13ab07d6a73385f5eda37aafc2dc31b#diff-fd2410931e7fdc4bf8b3ce23f5f7a27c7aacdf9337320626d86d806448c90b9bR1138

And simplified the tests so they don't rely on PutFile (also speeds things up a bit because we dont have to wait for filesystem io)

Reply:
Yes, good point: we used to treat a `Custom Text` set to empty the same as if it wasn't set at all, which is consistent with most other processors, and we probably don't want to change that.

I'm OK with either remembering in `onSchedule` if `Custom Text` was empty, and using that info here, or to just treat a `Custom Text` expanded to empty as if it was empty originally -- whichever you prefer.  A non-empty `Custom Text` expanding to empty is an edge case which may never happen in real life.

Reply:
I've refactored a little bit so Its easier to see what determines the output mode
Made sure that we dont break compatibility so CustomText "" is the same as no CustomText, but if CustomText were to evaluate to "" we would create an empty FF.
https://github.com/apache/nifi-minifi-cpp/pull/1706/commits/626d69c49f1c2ba04ccbd36d3ceba1fff6062e76

Reply:
Added an other test that covers this last use case https://github.com/apache/nifi-minifi-cpp/pull/1706/commits/13f0f74c7e46e4228c0b1cab7381a3b8416094ec